### PR TITLE
iSCSI UI

### PR DESCRIPTION
**Pull request by joseivanlopez**
## Problem

The storage client was already adapted for dealing with the iSCSI D-Bus API recently shipped by the storage sevice, see:

* https://github.com/yast/d-installer/pull/402
* https://github.com/yast/d-installer/pull/441  

But there is no UI yet for configuring iSCSI.

## Solution

Add a new page for configuring iSCSI.

## Testing

- Tested manually

NOTE: The UI was heavily tested in a manual way. Unit tests will come soon in a separate PR, once we have finished more urgent tasks.

## Screenshots

<details>
<summary>Show/Hide</summary>

Screenshot from 2023-03-16 13-20-06

Screenshot from 2023-03-16 13-20-19

Screenshot from 2023-03-16 13-20-34

Screenshot from 2023-03-16 13-20-52

Screenshot from 2023-03-16 13-21-25

</details>

### CODE REVIEW

File: web/src/components/storage/iscsi/InitiatorSection.jsx [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    loadInitiator().catch(console.error);
+
+    return client.iscsi.onInitiatorChanged(loadInitiator);
+  }, [cancellablePromise, client.iscsi]);
+
+  return (
+    <Section title="Initiator" iconName="settings">
+      <InitiatorPresenter
+        initiator={initiator}
+        client={client}
```
Comment (dgdavid) [inline web/src/components/storage/iscsi/InitiatorSection.jsx:51]: NP: I do not see the point of this props here. As far as I can see, the presenter receives it for passing down to the form. This reminds me the [props drilling issue](https://kentcdodds.com/blog/prop-drilling)

So, I would just grab the client directly in the form, like you did here. That's one of the reasons we have the `useInstallerClient`.

File: web/src/components/storage/iscsi/InitiatorPresenter.jsx [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+  );
+};
+
+export default function InitiatorPresenter({ initiator, client }) {
+  const [data, setData] = useState();
+  const [isFormOpen, setIsFormOpen] = useState(false);
+
+  useEffect(() => {
+    if (initiator !== undefined) setData({ ...initiator });
+  }, [setData, initiator]);
```
Comment (dgdavid) [inline web/src/components/storage/iscsi/InitiatorPresenter.jsx:57]: Actually this is not needed and might be causing an extra render. Simply initialize data with the prop value.

```suggestion
export default function InitiatorPresenter({ initiator, client }) {
  const [data, setData] = useState(initiator);
  const [isFormOpen, setIsFormOpen] = useState(false);
```

File: web/src/components/storage/iscsi/AuthFields.jsx [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+            style={{ verticalAlign: "text-bottom", marginRight: "0.3em" }}
+          />
+          Only available if authentication by target is provided
+        </p>
+      );
+    };
+
+    return (
+      <>
+        Authentication by inititator
```
Comment (dgdavid) [inline web/src/components/storage/iscsi/AuthFields.jsx:75]: typo: initi**t**ator

File: web/src/components/storage/iscsi/index.js [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+ *
+ * To contact SUSE LLC about this file by physical or electronic mail, you may
+ * find current contact information at www.suse.com.
+ */
+
+const NodeStartupOptions = Object.freeze({
+  MANUAL: { label: "Manual", value: "manual" },
+  ONBOOT: { label: "On boot", value: "onboot" },
+  AUTOMATIC: { label: "Automatic", value: "automatic" }
+});
```
Comment (dgdavid) [inline web/src/components/storage/iscsi/index.js:26]: NP: I'd move it to its own file, like the rest.

File: web/src/components/storage/iscsi/TargetsSection.jsx [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+  const openDiscoverForm = () => {
+    dispatch({ type: "OPEN_DISCOVER_FORM" });
+  };
+
+  const closeDiscoverForm = () => {
+    dispatch({ type: "CLOSE_DISCOVER_FORM" });
+  };
+
+  const SectionContent = () => {
+    if (state.isLoading) return <Skeleton />;
```
Comment (dgdavid) [inline web/src/components/storage/iscsi/TargetsSection.jsx:123]: Here I prefer `SectionSkeleton`. We are aware that it has to be polished, but please use it by now when you're waiting for the section content. The _raw_ Skeleton inside a row when waiting for its content is ok.

File: web/src/components/storage/iscsi/InitiatorSection.jsx (hunk 1 above)
Comment (joseivanlopez) [reply web/src/components/storage/iscsi/InitiatorSection.jsx:51]: I have improved the situation a bit. Now, the client is used by the sections in order to get the necessary data, and then they pass the data and the client to the presenters. The presenters still need a client to perform the dbus calls corresponding to each possible action in the table. In another iteratation we could avoid passing clients from sections to presenters. At least, now the presenter is not passed to the form components.

File: web/src/components/storage/iscsi/InitiatorForm.jsx [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+export default function InitiatorForm({ initiator, client, onSuccess, onCancel }) {
+  const [data, setData] = useState({ ...initiator });
+
+  const onNameChange = name => setData({ ...data, name });
+
+  const onSubmit = async (event) => {
+    event.preventDefault();
+    await client.iscsi.setInitiatorName(data.name);
+    onSuccess();
```
[the remark this answers is not in the document]
Comment (joseivanlopez) [reply web/src/components/storage/iscsi/InitiatorForm.jsx:35]: You are right. I have improved the forms. Thanks ;)

File: web/src/components/storage/iscsi/AuthFields.jsx [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+            type="password"
+            aria-label="Password"
+            value={data.password || ""}
+            label="Password"
+            onChange={onPasswordChange}
+            validated={showPasswordError() ? "error" : "default"}
+          />
+        </FormGroup>
+      </Fieldset>
+      <Fieldset legend={ByInitiatorAuthLegend()}>
```
Comment (dgdavid) [inline web/src/components/storage/iscsi/AuthFields.jsx:118]: NP: Not sure if it's a good idea adding the additional warning as part of the legend. But keep it as it is now, we can refine it in a later iteration.

File: web/src/components/storage/iscsi/InitiatorPresenter.jsx [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+  );
+
+  return (
+    <ActionsColumn
+      items={actions}
+      actionsToggle={actionsToggle}
+      {...props}
+    />
+  );
+};
```
Comment (dgdavid) [inline web/src/components/storage/iscsi/InitiatorPresenter.jsx:49]: Reminder: As discussed off-line, in future iterations we should extract this to a common component or similar. It's used here, at users and in another place I do not remember right now.

File: web/src/components/storage/iscsi/LoginForm.jsx [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, contact SUSE LLC.
+ *
+ * To contact SUSE LLC about this file by physical or electronic mail, you may
+ * find current contact information at www.suse.com.
+ */
+
+// cspell:ignore onboot
```
Comment (dgdavid) [inline web/src/components/storage/iscsi/LoginForm.jsx:22]: NP: as with startup, I think this word could be moved to the custom dictionary.

File: web/src/components/storage/iscsi/NodeStartupOptions.js [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, contact SUSE LLC.
+ *
+ * To contact SUSE LLC about this file by physical or electronic mail, you may
+ * find current contact information at www.suse.com.
+ */
+
+// cspell:ignore onboot
```
Comment (dgdavid) [inline web/src/components/storage/iscsi/NodeStartupOptions.js:22]: NP: same than above

File: web/src/index.js [hunk 11]
```diff
@@ -58,6 +58,7 @@ root.render(
                 <Route path="/storage" element={<StoragePage />} />
                 <Route path="/network" element={<NetworkPage />} />
                 <Route path="/users" element={<UsersPage />} />
+                <Route path="/storage/iscsi" element={<ISCSIPage />} />
```
Comment (dgdavid) [inline web/src/index.js:61]: NP: please, keep related routes close :stuck_out_tongue_closed_eyes:

File: web/src/utils.js [hunk 12]
```diff
@@ -143,9 +143,29 @@ function useCancellablePromise() {
...
+    JSON.parse(localStorage.getItem(storageKey)) ?? fallbackState
+  );
+
+  useEffect(() => {
+    localStorage.setItem(storageKey, JSON.stringify(value));
+  }, [value, storageKey]);
+
+  return [value, setValue];
+};
+
```
Comment (dgdavid) [inline web/src/utils.js:164]: :ok_hand:

**Review by dgdavid — APPROVED** [review]
LGTM

File: web/src/components/storage/iscsi/InitiatorPresenter.jsx (hunk 8 above)
Comment (joseivanlopez) [reply web/src/components/storage/iscsi/InitiatorPresenter.jsx:49]: It is implemeted by `ConnectionsTable`, `FirstUser`, `RootAuthMethods`, `InitiatorPresenter` and `NodesPresenter`.

File: web/src/client/storage.js [hunk 13]
```diff
@@ -296,23 +298,7 @@ class ISCSIManager {
...
-        interface: iscsiProxy.Interface,
-        ibft: iscsiProxy.IBFT,
-        connected: iscsiProxy.Connected,
-        startup: iscsiProxy.Startup
-      };
-    };
-
-    const proxy = await this.iscsiNodesProxy();
-    return Object.values(proxy).map(p => buildNode(p));
+    return Object.values(this.proxies.iscsiNodes).map(this.buildNode);
```
[the remark this answers is not in the document]
Comment (joseivanlopez) [reply web/src/client/storage.js:301]: Good point.

**Review by dgdavid — APPROVED** [review]
Looks good, thanks!

**Pull request merged**